app/controllers/rfid_controller.rb

```diff
@@ -26,25 +26,27 @@ def card_number
       end
     end
   end
-  
+
   def check_session(rfid)
     active_session = rfid.user.lab_sessions.where("sign_out_time > ?", Time.now)
     if active_session.present?
       active_session.update_all(sign_out_time: Time.now)
-      active_location = PiReader.find_by(pi_mac_address: active_session.first.try(:mac_address)).try(:pi_location)
-      new_location = PiReader.find_by(pi_mac_address: params[:mac_address]).try(:pi_location)
+      active_location = PiReader.find_by(pi_mac_address: active_session.first.try(:mac_address))
+      new_location = PiReader.find_by(pi_mac_address: params[:mac_address])
       if active_location != new_location
         new_session(rfid)
-      end   
+      end
     else
       new_session(rfid)
     end
   end
-  
+
   def new_session (rfid)
     sign_in = Time.now
     sign_out = sign_in + 3.hours
-    new_session = rfid.user.lab_sessions.new(sign_in_time: sign_in, sign_out_time: sign_out, mac_address: params[:mac_address])
-    new_session.save
+    if raspi = PiReader.find_by(pi_mac_address: params[:mac_address])
```

Comment:
Since you just found the PiReader in `check_session` you might as well pass it in as a parameter